common/flatpak-dir.c

```diff
@@ -8128,21 +8203,43 @@ flatpak_dir_find_remote_refs (FlatpakDir   *self,
   if (!flatpak_dir_ensure_repo (self, NULL, error))
     return NULL;
 
-  if (!flatpak_dir_remote_list_refs (self, remote,
-                                     &remote_refs, cancellable, error))
-    return NULL;
-
   collection_id = flatpak_dir_get_remote_collection_id (self, remote);
-  matched_refs = find_matching_refs (remote_refs,
-                                     name,
-                                     opt_branch,
-                                     opt_arch,
-                                     collection_id,
-                                     kinds,
-                                     FIND_MATCHING_REFS_FLAGS_NONE,
-                                     error);
-  if (matched_refs == NULL)
-    return NULL;
+  if (collection_id == NULL)
+    {
+      if (!flatpak_dir_remote_list_refs (self, remote,
+                                         &remote_refs, cancellable, error))
+        return NULL;
+
+      matched_refs = find_matching_refs (remote_refs,
+                                         name,
+                                         opt_branch,
+                                         opt_arch,
+                                         NULL,
+                                         kinds,
+                                         FIND_MATCHING_REFS_FLAGS_NONE,
+                                         error);
+      if (matched_refs == NULL)
+        return NULL;
+    }
+  else
+    {
+      g_autoptr(GVariant) xa_cache = NULL;
+
+      /* Use xa.cache because we can't fetch the summary when offline */
+      if (!flatpak_dir_fetch_refs_cache (self, remote, &xa_cache, cancellable, error))
+        return NULL;
+
+      matched_refs = find_matching_refs_from_cache (xa_cache,
```

Comment:
Is there a way you could merge `find_matching_refs()` and `find_matching_refs_from_cache()`? These two branches (for `collection == NULL`) are really quite similar, and I suspect they’re going to have to be replicated wherever `find_matching_refs()` is called.

Alternatively, you could factor out this whole `if (collection_id == NULL) … else …` block into a function which returns `matched_refs`.

Reply:
Not really possible to merge them, because they're iterating over different data structures. I suppose you could have the caller pass one or the other (the xa.cache or the summary), but that seems ugly. I don't think factoring out the whole block is necessary either because this is the only place it's used.